plutus-pab/tx-inject/Main.hs

```diff
@@ -0,0 +1,214 @@
+import           Control.Lens                   hiding (ix)
+import           Control.Monad                  (forever)
+import           Control.Monad.IO.Class         (liftIO)
+import           Control.RateLimit              (rateLimitExecution)
+import qualified Data.Map                       as Map
+import           Data.Text                      (Text)
+import qualified Data.Text                      as T
+import qualified Data.Text.IO                   as T
+import           Data.Time.Units                (Microsecond, fromMicroseconds)
+import           Data.Yaml                      (decodeFileThrow)
+import           GHC.Generics                   (Generic)
+import           Network.HTTP.Client            (ManagerSettings (..), newManager, setRequestIgnoreStatus)
+import           Network.HTTP.Conduit           (tlsManagerSettings)
+import           Options.Applicative            (Parser, ParserInfo, auto, execParser, fullDesc, help, helper, info,
+                                                 long, metavar, option, progDesc, short, showDefault, strOption, value,
+                                                 (<**>))
+import           Servant.Client                 (ClientEnv (..), mkClientEnv, runClientM)
+import           System.Clock                   (Clock (..), TimeSpec (..), getTime)
+import           System.Random.MWC              (GenIO, createSystemRandom)
+import           System.Signal                  (installHandler, sigINT)
+import           Text.Pretty.Simple             (pPrint)
+
+import           Cardano.Node.Client            (addTx)
+import           Cardano.Node.RandomTx          (generateTx)
+import           Cardano.Node.Server            (MockServerConfig (..))
+import           Language.Plutus.Contract.Trace (defaultDist)
+import           Ledger.Index                   (UtxoIndex (..), insertBlock)
+import           Ledger.Tx                      (Tx (..))
+import           Plutus.PAB.Types               (Config (..))
+import           Wallet.Emulator                (chainState, txPool, walletPubKey)
+import           Wallet.Emulator.MultiAgent     (emulatorStateInitialDist)
+
+{- | The `Stats` are used by both the producer and consumer to track the number of
+     generated transactions (used to verify if we respect the requested TPS rate)
+     and the number of UTxOs (not used at the moment)
+-}
+data Stats = Stats
+  { stStartTime :: TimeSpec
+  , stCount     :: Integer
+  , stUtxoSize  :: Int
+  , stEndTime   :: TimeSpec
+  } deriving (Show, Generic)
+
+{- | The `AppEnv` is used by both the producer and the consumer to pass the data
+     required for execution
+-}
+data AppEnv = AppEnv
+  { clientEnv :: ClientEnv
+  , txQueue   :: TBQueue Tx
+  , stats     :: TVar Stats
+  , utxoIndex :: UtxoIndex
+  }
+
+{- | This builds the default UTxO index, using 10 wallets. We need to use the same
+     number of wallets and addresses as the PAB such that the generated data will
+     be validated by the PAB
```

Comment:
transactions are validated by the node, not the PAB. And hypothetically, if we used different addresses here from the ones we use in the PAB, the PAB wouldn't care (or at least it _shouldn't_ care)

Reply:
I will drop the validated by the PAB bit, but I think we still want the same addresses to get a consistent view of the utxo set, for example. Am I wrong?

Reply:
It could also be that the tx generator has a subset of the UTXO set completely disjoint from the subset that the PAB is interested in. There is no inherent reason why they need to share the same addresses.

Reply:
I removed the reference to synchronising state.